### PR TITLE
Check PHP8 + add Phpstan + some improvment

### PR DESCRIPTION
**Pull request by Benoit382**
- Check compatibility with PHP8
- Update PHPunit
- Add (PHPSstan)[https://github.com/phpstan/phpstan]
- Add cs-fixer with local rule (feel free to update)
- Improve some test (too many copy paste)
- Say "What fucking code" and gitblame myself... 

Ready for all comment|blame ;)

### CODE REVIEW

**Comment by gherardobertini** [comment]
Have you tried to add `php 8.0` in travis?  
I notice that I can't install the package because is stuck to `psr/log ^1.0`

**Comment by Dwarfex** [comment]
You could merge the PR https://github.com/openfoodfacts/openfoodfacts-php/pull/32 into your branch, and check if the Github Actions will run successfull.

File: README.md [hunk 1]
```diff
@@ -68,11 +68,11 @@ see: [PSR-16 Simple Cache](https://www.php-fig.org/psr/psr-16/)
...
 
 1. Fork it ( https://github.com/openfoodfacts/openfoodfacts-php/fork )
 2. Create your feature branch (`git checkout -b my-new-feature`)
-3. Commit your changes (`git commit -am 'Add some feature'`)
-4. Push to the branch (`git push origin my-new-feature`)
-5. Create a new Pull Request
+3. Launch test `vendor/bin/phpunit` && cs-fixer `vendor/bin/php-cs-fixer fix`
+4. Commit your changes (`git commit -am 'Add some feature'`)
+5. Push to the branch (`git push origin my-new-feature`)
+6. Create a new Pull Request
```
Comment (Dwarfex) [inline README.md:78]: 7. Check that all Tests in the PR run

File: composer.json [hunk 2]
```diff
@@ -20,22 +21,26 @@
...
-    "psr/simple-cache": "^1.0",
-    "ext-json": "*",
-    "ext-curl": "*"
+    "psr/simple-cache": "^1.0"
   },
   "autoload": {
     "psr-4": {
-      "OpenFoodFacts\\": "src/"
+      "OpenFoodFacts\\": "src/",
+      "Tests\\": "tests/"
```
Comment (Dwarfex) [inline composer.json]: I would use a more distinct namespace than "Tests" - more like "OpenFoodFactsTests"

File: src/Collection.php [hunk 3]
```diff
@@ -2,15 +2,23 @@
...
+    /** @var array<int, Document> */
+    private $listDocuments  = [];
+    /** @var int */
+    private $count          = 0;
+    /** @var int */
+    private $page           = 0;
+    /** @var int */
+    private $skip           = 0;
+    /** @var int */
+    private $pageSize       = 0;
```
Comment (Dwarfex) [inline src/Collection.php:21]: I would suggest this:
```suggestion
    /** @var array<int, Document> */
    private array $listDocuments  = [];
    private int $count          = 0;
    private int $page           = 0;
    private int $skip           = 0;
    private int $pageSize       = 0;
```

File: src/Collection.php [hunk 4]
```diff
@@ -101,20 +109,23 @@ public function rewind()
...
      */
     public function current()
     {
         return current($this->listDocuments);
     }
     /**
      * @inheritDoc
+     * @return int|null
      */
     public function key()
```
Comment (Dwarfex) [inline src/Collection.php:122]:
```suggestion
     */
    public function key(): ?int
```

**Review by Dwarfex — CHANGES_REQUESTED** [review]
Let's disscuss the suggested changes if needed.

File: src/Collection.php (hunk 3 above)
Comment (Benoit382) [reply src/Collection.php:21]: The typed properties is not compatible with PHP 7.1
https://stitcher.io/blog/typed-properties-in-php-74

File: src/Collection.php (hunk 4 above)
Comment (Benoit382) [reply src/Collection.php:122]: The typed properties is not compatible with PHP 7.1
https://stitcher.io/blog/typed-properties-in-php-74

File: README.md (hunk 1 above)
Comment (Benoit382) [reply README.md:78]: Last time,  all test pass but today, some one didn't want to pass...
I think we should mock the openfoodfact response.

**Comment by Benoit382** [comment]
> Have you tried to add `php 8.0` in travis? I notice that I can't install the package because is stuck to `psr/log ^1.0`

I will check that.

**Comment by Benoit382** [comment]
I pick all code from #32

File: README.md (hunk 1 above)
Comment (Dwarfex) [reply README.md:78]: Yeah, that would be best. 
Testing the SKD and indirectly testing the API for correct behaviour is not best practice.

File: src/Collection.php (hunk 3 above)
Comment (Dwarfex) [reply src/Collection.php:21]: Ah - the fate of old code :D - we then might want to up the requirements to 7.3/7.4 (https://www.php.net/supported-versions.php) and add a new semantic release.

**Review by Dwarfex — APPROVED** [review]

**Comment by Benoit382** [comment]
I think we should squash and merge, and create a new issues, with bump version + create Unit|Functional test and remove integration test

**Comment by Dwarfex** [comment]
@Benoit382 that works for me :)

**Comment by Benoit382** [comment]
Some one can squash and merge ?

**Pull request merged**